scarlet2/measure.py

```diff
@@ -124,71 +126,215 @@ def moments(component, N=2, center=None, weight=None):
-                center = center[0][:, None, None], center[1][:, None, None]
-            grid_y = grid_y - center[0]
-            grid_x = grid_x - center[1]
-
-    return M
-
-
-def size(moments):
-    """Determine size from moments
-
-    Parameters
-    ----------
-    moments: moment dictionary from moments()
-    """
-    flux = moments[0, 0]
-    T = (moments[0, 2] / flux * moments[2, 0] / flux - (moments[1, 1] / flux) ** 2) ** (1 / 4)
-    return T
-
-
-def ellipticity(moments):
-    """Determine complex ellipticity from moments.
-
-    Parameters
-    ----------
-    moments: moment dictionary from moments()
-
-    Returns:
-    --------
-    jnp.array
-    """
-    ellipticity = (moments[0, 2] - moments[2, 0] + 2j * moments[1, 1]) / (moments[2, 0] + moments[0, 2])
-    return jnp.array((ellipticity.real, ellipticity.imag))
+    return Moments(component, N=N, center=center, weight=weight)
+
+
+class Moments(dict):
+
+    def __init__(self, component, N=2, center=None, weight=None):
+        """Compute moments of the light distribution.
+
+         Parameters
+         ----------
+         component: `scarlet.Component` or array
+             Component to analyze or its hyperspectral model
+         N: int >=0
+             Moment order
+         center: array
+             2D coordinate in frame of `component`
+         weight: array
+             weight function with same shape as `component`
+         """
+        super().__init__()
+
+        if isinstance(component, Component):
+            model = component()
+        else:
+            model = component
+
+        if weight is None:
+            weight = 1
+        self.weight = weight
+
+        self.model = model
+        grid_y, grid_x = jnp.indices(model.shape[-2:])
+
+        if model.ndim == 3:
+            grid_y = grid_y[None, :, :]
+            grid_x = grid_x[None, :, :]
+
+        self.N = N
+        for n in range(self.N + 1):
+            for m in range(n + 1):
+                # moments ordered by power in y, then x
+                self[m, n - m] = (grid_y ** m * grid_x ** (n - m) * model * weight).sum(axis=(-2, -1))
+
+            if n == 1:
+                # shift grid to produce centered moments
+                if center is None:
+                    center = self[1, 0] / self[0, 0], self[0, 1] / self[0, 0]  # centroid
+                    self[1, 0] = jnp.zeros_like(self[1, 0])
+                    self[0, 1] = jnp.zeros_like(self[0, 1])
+                else:
+                    center = jnp.asarray(center)
+                    # centroid wrt given center
+                    self[1, 0] -= center[0] * self[0, 0]
+                    self[0, 1] -= center[1] * self[0, 0]
+                if model.ndim == 3 and center[0].ndim == 1:
+                    center = center[0][:, None, None], center[1][:, None, None]
+                grid_y = grid_y - center[0]
+                grid_x = grid_x - center[1]
+    
+    def higher_order_moments(self):
```

Comment:
What is this function for? From the facto that it goes from moment N to 2N, I'm guessing to get the variances...

However, the constructor can measure any moment you want, but this code requires the model and the weights to be saved, which I am not a fan of because there's ample change to modify one without the others updating.

Reply:
In order to update moments from 0 to N, we also need to use moments from N+1 to 2N (see [Teague (1980), eq. 36](https://opg.optica.org/josa/fulltext.cfm?uri=josa-70-8-920&id=57703)), so I proposed to compute these additional moments in another function to update the first N moments stored the class. I am not a fan either that it requires to have access to the model. The problem is that once we updated the first N moments, the moments up to 2N are outdated since still correspond to the untransformed original image, so cannot be used for another transformation... 

Unless we keep track of the transformations in the class, and if the object is affected by another transformation we know the corresponding higher moments of the original image. What do you think?

Reply:
I don't see what you mean. Teague's eq. 36 clearly shows that you don't need higher moments. The paper even clarifies this point in the text below.

Reply:
This still is an open issue. I would much prefer the `Moment` instances not holding on to the data they were computed with.

Reply:
I got confused due to the j+k-r-s index, which I thought could go up to 4 if j and k are both equal to 2. But the sum is never above 2 for 2nd order moments (j=m, k=n-m)

Reply:
This should be fixed now :-)